crates/compiler/mono/src/inc_dec.rs

```diff
@@ -1,1554 +1,1193 @@
-        l: InLayout<'a>,
-        b: &'a Stmt<'a>,
-        b_live_vars: &LiveVarSet,
-    ) -> &'a Stmt<'a> {
-        use crate::low_level::HigherOrder::*;
-        use DataFunction::*;
-
-        let HigherOrderLowLevel {
-            op,
-            passed_function,
-            ..
-        } = lowlevel;
-
-        macro_rules! create_call {
-            ($borrows:expr) => {
-                create_holl_call(self.arena, lowlevel, $borrows, arguments)
-            };
+            triples
+                .into_iter()
+                .rev()
+                // First evaluate the continuation and let it consume it's free symbols.
+                .fold(
+                    insert_refcount_operations_stmt(arena, environment, current_stmt),
+                    |new_stmt, (binding, expr, layout)| {
+                        // If the binding is still owned in the environment, it is not used in the continuation and we can drop it right away.
+                        let new_stmt_without_unused = match environment
+                            .get_symbol_ownership(binding)
+                        {
+                            Some(Ownership::Owned) => insert_dec_stmt(arena, *binding, new_stmt),
+                            _ => new_stmt,
+                        };
+
+                        // And as the symbol should not be in scope before this let binding, remove it from the environment.
+                        environment.remove_symbol(*binding);
+
+                        insert_refcount_operations_binding(
+                            arena,
+                            environment,
+                            binding,
+                            expr,
+                            layout,
+                            new_stmt_without_unused,
+                        )
+                    },
+                )
         }
+        Stmt::Switch {
+            cond_symbol,
+            cond_layout,
+            branches,
+            default_branch,
+            ret_layout,
+        } => {
+            let new_branches = branches
+                .iter()
+                .map(|(label, info, branch)| {
+                    let mut branch_env = environment.clone();
 
-        let function_layout = ProcLayout {
-            arguments: passed_function.argument_layouts,
-            result: passed_function.return_layout,
-            niche: passed_function.name.niche(),
-        };
+                    let new_branch =
+                        insert_refcount_operations_stmt(arena, &mut branch_env, branch);
 
-        let function_ps = match self.param_map.get_symbol(
-            self.layout_interner,
-            passed_function.name.name(),
-            function_layout,
-        ) {
-            Some(function_ps) => function_ps,
-            None => unreachable!(),
-        };
+                    (*label, info.clone(), new_branch, branch_env)
+                })
+                .collect::<std::vec::Vec<_>>();
```

Comment:
you can use the `collect_in` function from `bumpalo`s `CollectIn` trait to create a bumpalo `Vec<_>` here. that is slightly more efficient.